File: ios/FamilyApp/Features/Family/FamilyViewModel.swift
// Family view model — the iOS twin of FamilyViewModel.kt.
import Foundation
import Observation
import Supabase

private let joinCodeLength = 8

@Observable
@MainActor
final class FamilyViewModel {
    private(set) var family: FamilyModel?
    private(set) var members: [UserModel] = []
    private(set) var currentUser: UserModel?
    /// My relation to each other member (their id → relation label), from my perspective.
    private(set) var relations: [String: String] = [:]
    var error: String?
    private(set) var isUploading = false

    /// Invite code captured from a deep link; FamilyScreen opens the join flow with it.
    var pendingJoinCode: String? {
        repo.pendingJoinCode
    }

    func consumePendingJoinCode() -> String? {
        repo.consumePendingJoinCode()
    }

    private let repo = FamilyRepository.shared
    private var client: SupabaseClient {
        SupabaseClientProvider.client
    }

    init() {
        refresh()
    }

    func refresh() {
        Task { await load() }
    }

    private func load() async {
        guard let userId = repo.session.currentUserId else {
            family = nil
            members = []
            currentUser = nil
            return
        }
        guard let user = await repo.getUser(userId) else { return }
        currentUser = user
        if let familyId = user.familyId {
            family = await repo.getFamily(familyId: familyId)
            members = await repo.getFamilyMembers(familyId: familyId)
            relations = await repo.getMyRelations(userId: userId)
        } else {
            family = nil
            members = []
            relations = [:]
        }
    }

    /// Sets my relation ("Dad", "Wife", …) to another member, or clears it when empty.
    func setRelation(toUserId: String, relation: String) {
        guard let userId = repo.session.currentUserId, let familyId = family?.id else { return }
        // Optimistic local update.
        if relation.trimmingCharacters(in: .whitespaces).isEmpty {
            relations.removeValue(forKey: toUserId)
        } else {
            relations[toUserId] = relation
        }
        Task {
            await repo.setRelation(
                fromUserId: userId, toUserId: toUserId, familyId: familyId, relation: relation
            )
        }
    }

    func clearError() {
        error = nil
    }

    func createFamily(name: String, code: String) {
        Task {
            guard let userId = repo.session.currentUserId else { return }
            do {
                _ = try await repo.createFamily(name: name, code: code, userId: userId)
                await load()
            } catch {
                self.error = error.localizedDescription
            }
        }
    }

    /// Set right after joining a family with existing members — FamilyScreen shows the
    /// "set your relations" prompt so the newcomer can label the people already there.
    var promptRelationSetup = false

    func joinFamily(code: String) {
        Task {
            guard let userId = repo.session.currentUserId else { return }
            do {
                _ = try await repo.joinFamily(code: code, userId: userId)
                await load()
                // Offer to set relations to the members already in the family.
                if members.contains(where: { $0.id != userId }) {
                    promptRelationSetup = true
                }
            } catch {
                self.error = (error as? RepositoryError)?.errorDescription ?? error.localizedDescription
            }
        }
    }

    func leaveFamily() {
        Task {
            guard let userId = repo.session.currentUserId else { return }
            await repo.leaveFamily(userId: userId)
            await load()
        }
    }

    func renameFamily(_ newName: String) {
        Task {
            guard let familyId = family?.id else { return }
            do {
                try await repo.renameFamily(familyId: familyId, newName: newName)
                await load()
            } catch {
                self.error = error.localizedDescription
            }
        }
    }

    func uploadFamilyPhoto(_ imageData: Data) {
        Task {
            guard let familyId = family?.id else { return }
            isUploading = true
            defer { isUploading = false }
            guard let compressed = ImageUtils.compressWithOrientation(imageData) else {
                error = L("Could not read the selected photo.")
                return
            }
            do {
                // Same bucket + path convention as Android: group-images/family-photos/{id}/photo.jpg
                let bucket = client.storage.from("group-images")
                let path = "family-photos/\(familyId)/photo.jpg"
                try await bucket.upload(path, data: compressed, options: FileOptions(upsert: true))
                let cacheBuster = Int(Date().timeIntervalSince1970 * 1000)
                let url = try bucket.getPublicURL(path: path).absoluteString + "?t=\(cacheBuster)"
                try await repo.updateFamilyPhoto(familyId: familyId, photoUrl: url)
                await load()
            } catch {
                self.error = L("Failed to update photo. Please try again.")
            }
        }
    }

    func removeMember(_ memberId: String) {
        Task {
            do {
                try await repo.removeFamilyMember(memberId: memberId)
                await load()
            } catch {
                self.error = error.localizedDescription
            }
        }
    }
}

/// 8-char uppercase invite code — mirrors generateJoinCode in FamilyViewModel.kt.
func generateJoinCode() -> String {
    String(UUID().uuidString.prefix(joinCodeLength)).uppercased()
}

/// Share-sheet invite message — mirrors the Android share text. Pass `appLocale` so
/// the chrome around the (verbatim) family name and code renders in the app language.
func inviteMessage(
    familyName: String,
    joinCode: String,
    locale: Locale = Locale(identifier: "en_US_POSIX")
) -> String {
    let link = DeepLinkURL.invite(code: joinCode).absoluteString
    return L(
        """
        Join our family "\(familyName)" on The Family App!

        Tap to join: \(link)

        Or open the app, tap "Join with Invite Code", and enter: \(joinCode)
        """,
        locale: locale
    )
}

File: android/app/src/main/java/com/sandnes/familyapp/ui/family/FamilyViewModel.kt
package com.sandnes.familyapp.ui.family

import android.app.Application
import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sandnes.familyapp.R
import com.sandnes.familyapp.data.FamilyModel
import com.sandnes.familyapp.data.FamilyRepository
import com.sandnes.familyapp.data.UserModel
import com.sandnes.familyapp.data.getFamilyRelations
import com.sandnes.familyapp.data.remote.SupabaseManager
import com.sandnes.familyapp.data.setFamilyRelation
import com.sandnes.familyapp.util.compressImageWithOrientation
import dagger.hilt.android.lifecycle.HiltViewModel
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.storage.storage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

private const val JOIN_CODE_LENGTH = 8
private const val MAX_IMAGE_DIM = 1024
private const val JPEG_QUALITY = 85

@HiltViewModel
class FamilyViewModel
    @Inject
    constructor(
        private val app: Application,
        internal val repo: FamilyRepository,
    ) : ViewModel() {
        private val _family = MutableStateFlow<FamilyModel?>(null)
        val family: StateFlow<FamilyModel?> = _family.asStateFlow()

        private val _members = MutableStateFlow<List<UserModel>>(emptyList())
        val members: StateFlow<List<UserModel>> = _members.asStateFlow()

        private val _currentUser = MutableStateFlow<UserModel?>(null)
        val currentUser: StateFlow<UserModel?> = _currentUser.asStateFlow()

        private val _error = MutableStateFlow<String?>(null)
        val error: StateFlow<String?> = _error.asStateFlow()

        private val _isUploading = MutableStateFlow(false)
        val isUploading: StateFlow<Boolean> = _isUploading.asStateFlow()

        /** My relation to each other member (their id → relation label), from my perspective. */
        private val _relations = MutableStateFlow<Map<String, String>>(emptyMap())
        val relations: StateFlow<Map<String, String>> = _relations.asStateFlow()

        /**
         * Set right after joining a family that already has members, so the screen can prompt the
         * newcomer to label the people already there. Mirrors iOS `promptRelationSetup`.
         */
        private val _promptRelationSetup = MutableStateFlow(false)
        val promptRelationSetup: StateFlow<Boolean> = _promptRelationSetup.asStateFlow()

        fun dismissRelationSetup() {
            _promptRelationSetup.value = false
        }

        /** Invite code captured from a deep link; FamilyScreen opens the join flow with it. */
        val pendingJoinCode = repo.pendingJoinCode

        fun consumePendingJoinCode() = repo.consumePendingJoinCode()

        init {
            viewModelScope.launch {
                repo.currentUserId.collect { userId ->
                    if (userId != null) {
                        load(userId)
                    } else {
                        _family.value = null
                        _members.value = emptyList()
                        _currentUser.value = null
                    }
                }
            }
        }

        fun refresh() =
            viewModelScope.launch {
                val userId = repo.currentUserId.first() ?: return@launch
                load(userId)
            }

        private suspend fun load(userId: String) {
            runCatching {
                val user = repo.getUser(userId) ?: return
                _currentUser.value = user
                if (user.familyId != null) {
                    _family.value = repo.getFamily(user.familyId)
                    _members.value =
                        SupabaseManager.client.postgrest
                            .from("users")
                            .select { filter { eq("family_id", user.familyId) } }
                            .decodeList<UserModel>()
                    _relations.value =
                        repo
                            .getFamilyRelations(user.familyId)
                            .filter { it.fromUserId == userId }
                            .associate { it.toUserId to it.relation }
                } else {
                    _family.value = null
                    _members.value = emptyList()
                    _relations.value = emptyMap()
                }
            }
        }

        /**
         * Sets my relation ("Dad", "Wife", …) to another member, or clears it when blank. Optimistic
         * local update, then persists (RLS enforces from_user_id = me). Mirrors iOS `setRelation`.
         */
        fun setRelation(
            toUserId: String,
            relation: String,
        ) = viewModelScope.launch {
            val userId = repo.currentUserId.first() ?: return@launch
            val familyId = _family.value?.id ?: return@launch
            _relations.update { current ->
                if (relation.isBlank()) current - toUserId else current + (toUserId to relation)
            }
            repo.setFamilyRelation(userId, toUserId, familyId, relation)
        }

        fun clearError() {
            _error.value = null
        }

        fun createFamily(
            name: String,
            code: String,
        ) =
            viewModelScope.launch {
                val userId = repo.currentUserId.first() ?: return@launch
                repo
                    .createFamily(name, code, userId)
                    .onSuccess { load(userId) }
                    .onFailure { _error.value = it.message }
            }

        fun joinFamily(code: String) =
            viewModelScope.launch {
                val userId = repo.currentUserId.first() ?: return@launch
                repo
                    .joinFamily(code, userId)
                    .onSuccess {
                        load(userId)
                        // Offer to set relations to the members already in the family.
                        if (_members.value.any { it.id != userId }) {
                            _promptRelationSetup.value = true
                        }
                    }.onFailure { _error.value = it.message }
            }

        fun leaveFamily() =
            viewModelScope.launch {
                val userId = repo.currentUserId.first() ?: return@launch
                repo.leaveFamily(userId)
                load(userId)
            }

        fun generateJoinCode(): String =
            java.util.UUID
                .randomUUID()
                .toString()
                .take(JOIN_CODE_LENGTH)
                .uppercase()

        fun renameFamily(newName: String) {
            val fid = _family.value?.id ?: return
            viewModelScope.launch {
                repo
                    .renameFamily(fid, newName)
                    .onSuccess {
                        val userId = repo.currentUserId.first() ?: return@launch
                        load(userId)
                    }.onFailure { _error.value = it.message }
            }
        }

        fun uploadFamilyPhoto(
            context: Context,
            uri: Uri,
        ) {
            val fid = _family.value?.id ?: return
            viewModelScope.launch {
                _isUploading.value = true
                try {
                    runCatching {
                        val raw =
                            withContext(Dispatchers.IO) {
                                context.contentResolver.openInputStream(uri)?.use { it.readBytes() }
                            } ?: run {
                                _error.value = app.getString(R.string.could_not_read_the_selected_photo)
                                return@launch
                            }
                        val compressed = compressImage(raw)
                        val bucket = SupabaseManager.client.storage.from("group-images")
                        val path = "family-photos/$fid/photo.jpg"
                        bucket.upload(path, compressed) { upsert = true }
                        val url = bucket.publicUrl(path) + "?t=${System.currentTimeMillis()}"
                        repo.updateFamilyPhoto(fid, url).getOrThrow()
                        val userId = repo.currentUserId.first() ?: return@launch
                        load(userId)
                    }.onFailure { e ->
                        Log.e("FamilyVM", "Photo upload failed", e)
                        _error.value = app.getString(R.string.failed_to_update_photo_please_try_again)
                    }
                } finally {
                    _isUploading.value = false
                }
            }
        }

        private suspend fun compressImage(bytes: ByteArray): ByteArray =
            withContext(Dispatchers.IO) { compressImageWithOrientation(bytes, MAX_IMAGE_DIM, JPEG_QUALITY) }

        fun removeMember(memberId: String) {
            viewModelScope.launch {
                repo
                    .removeFamilyMember(memberId)
                    .onSuccess {
                        val userId = repo.currentUserId.first() ?: return@launch
                        load(userId)
                    }.onFailure { _error.value = it.message }
            }
        }
    }
